File: api/analyze.js
const { GoogleGenAI } = require('@google/genai');

module.exports = async (req, res) => {
    // Jalur proteksi metode tembakan request
    if (req.method !== 'POST') {
        return res.status(405).json({ error: 'Method tidak diizinkan, Bos!' });
    }

    try {
        const { image } = req.body;
        if (!image) {
            return res.status(400).json({ error: 'Gambar Base64 tidak dimuat, Bos!' });
        }

        // Mengambil API Key rahasia dari environment variable Vercel Bos
        const apiKey = process.env.GEMINI_API_KEY;
        if (!apiKey) {
            return res.status(500).json({ error: 'Kunci rahasia GEMINI_API_KEY belum dipasang di dashboard Vercel, Bos!' });
        }

        const ai = new GoogleGenAI({ apiKey: apiKey });

        // Perintah system prompt agar Gemini mengekstrak komponen gambar menjadi prompt matang
        const promptInstruction = "Analyze this image and extract its components to generate a highly detailed prompt for image generation. Include aspect ratio, style, lighting, camera lens description, textures, colors, and clothing details. Output ONLY the prompt in English text, no greetings or explanations.";

        const response = await ai.models.generateContent({
            model: 'gemini-2.5-flash',
            contents: [
                promptInstruction,
                {
                    inlineData: {
                        mimeType: 'image/jpeg',
                        data: image
                    }
                }
            ]
        });

        // Kembalikan teks prompt matang ke frontend index.html
        return res.status(200).json({ text: response.text });

    } catch (error) {
        return res.status(500).json({ error: 'API Gemini sedang padat lintasan, Bos!' });
    }
};

File: api/analyze.py
import os
import json
import urllib.request
import urllib.parse

def handler(request):
    # Hanya menerima metode POST kiriman foto dari frontend
    if request.method != 'POST':
        return {
            'statusCode': 405,
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps({'error': 'Metode tidak diizinkan'})
        }

    try:
        # Membaca body request foto base64
        req_body = json.loads(request.body.decode('utf-8'))
        image_base64 = req_body.get("image", "")
        
        # Mengambil API Key rahasia dari Settings Vercel
        api_key = os.environ.get("GEMINI_API_KEY")
        
        if not api_key:
            return {
                'statusCode': 500,
                'headers': {'Content-Type': 'application/json'},
                'body': json.dumps({'error': 'API Key Gemini belum diisi di Environment Variables Vercel Bos!'})
            }

        # Setup URL Google Gemini API 1.5 Flash (Sangat Cepat & Akurat)
        url = f"https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key={api_key}"
        
        instruksi_analisis = (
            "Analisis foto referensi ini secara mendalam untuk kebutuhan profesional AI Image Generation Prompt (seperti Midjourney, Stable Diffusion, atau Gemini App).\n\n"
            "=== ATURAN UTAMA (WAJIB & MUTLAK) ===\n"
            "1. JAGA WAJAH ASLI: Saat membuat output prompt, instruksikan AI generator untuk MENGGUNAKAN WAJAH ASLI dari foto referensi ini secara TOTAL.\n"
            "2. TANPA MODIFIKASI WAJAH: Wajah harus tetap realistis, identik, dan fotorealistik dengan foto asli. Dilarang mengubah bentuk wajah, fitur wajah, atau ekspresi dasar yang membuatnya tidak dikenali.\n"
            "3. Identitas wajah harus 100% sama dengan input.\n\n"
            "=== FORMAT OUTPUT BAHASA INDONESIA ===\n"
            "Berikan hasil analisis dalam Bahasa Indonesia yang rapi dengan poin-poin berikut:\n"
            "- Subjek Utama\n"
            "- Pakaian & Atribut\n"
            "- Latar Belakang/Suasana\n"
            "- Pencahayaan & Gaya Kamera\n"
            "- Postur Tubuh (Detail, tinggi, pose)\n\n"
            "=== FORMAT PROMPT ENGLISH (DI PALING BAWAH) ===\n"
            "Tuliskan kata 'PROMPT ENGLISH:' (huruf besar semua), lalu buatkan satu paragraf prompt English yang sangat detail, fotorealistik, dan hyper-realistic.\n\n"
            "Suntikkan perintah English di dalam prompt tersebut (seperti: 'utilizing the exact, unmodified, highly detailed face from the reference image', 'photorealistic identity match', 'the face remains completely identical to the source, without any AI alteration or smoothing')."
        )

        gemini_payload = {
            "contents": [{
                "parts": [
                    {"text": instruksi_analisis},
                    {
                        "inlineData": {
                            "mimeType": "image/jpeg",
                            "data": image_base64
                        }
                    }
                ]
            }],
            "generationConfig": {
                "temperature": 0.3,
                "topP": 0.8,
                "topK": 40
            }
        }

        # Hantam data ke server Google API
        req = urllib.request.Request(
            url,
            data=json.dumps(gemini_payload).encode('utf-8'),
            headers={'Content-Type': 'application/json'},
            method='POST'
        )

        with urllib.request.urlopen(req) as response:
            res_data = json.loads(response.read().decode('utf-8'))
            extracted_text = res_data['candidates'][0]['content']['parts'][0]['text']
            
            return {
                'statusCode': 200,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*'
                },
                'body': json.dumps({'text': extracted_text})
            }

    except Exception as e:
        return {
            'statusCode': 500,
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps({'error': f'Sistem Crash: {str(e)}'})
        }
